compiler/hash-typecheck/src/new/ops/normalise.rs

```diff
@@ -0,0 +1,31 @@
+use derive_more::Constructor;
+use hash_types::new::terms::TermId;
+
+use crate::{
+    impl_access_to_tc_env,
+    new::{diagnostics::error::TcResult, environment::tc_env::TcEnv},
+};
+
+#[derive(Constructor)]
+pub struct NormalisationOps<'tc> {
+    tc_env: &'tc TcEnv<'tc>,
+}
+
+impl_access_to_tc_env!(NormalisationOps<'tc>);
+
+impl<'tc> NormalisationOps<'tc> {
+    /// Turn a term to weak head normal form.
+    pub fn standardise_term(&self, _term_id: TermId) -> TcResult<TermId> {
+        todo!()
+    }
+
+    /// Turn a term to weak head normal form, or return `None` .
+    pub fn potentially_standardise_term(&self, _term_id: TermId) -> TcResult<Option<TermId>> {
+        todo!()
+    }
+
+    /// Reduce a term to normal form.
+    pub fn potentially_normalise_term(&self, _term_id: TermId) -> TcResult<Option<TermId>> {
```

Comment:
So I'm guessing this will create a new term rather than modify in-place? I am a little weary of this approach and I think it should be discussed a bit on how to make it more memory compact.

Reply:
I also expect that most benefits will come from filling holes in place rather than duplicating the term with the filled hole, but again there are sometimes situations where you have to copy.

Reply:
Yeah the old implementation is very liberal with how many terms it creates. I think for this implementation it is best to wait and see how it runs and then optimise as needed/where is most beneficial.